xmodule/capa/safe_exec/safe_exec.py

```diff
@@ -21,7 +21,17 @@
 from __future__ import absolute_import, division
 
 import os
-os.environ["OPENBLAS_NUM_THREADS"] = "1"    # See TNL-6456
+
+# openblas is a math library used by numpy. It will try to allocate multiple
+# threads by default, but this may exceed resource limits and cause a segfault.
+# Limiting to 1 thread will prevent this in all configurations.
+os.environ["OPENBLAS_NUM_THREADS"] = "1"
+
+# Any code that uses the tempfile module to create temporary files should use
+# the ./tmp directory that codejail creates in each sandbox, rather than trying
+# to use a global temp dir (which should be blocked by AppArmor anyhow).
+# This is needed for matplotlib among other things.
+os.environ["TMPDIR"] = os.getcwd() + "/tmp"
 
 import random2 as random_module
```

Comment:
While I was testing this i noticed the following error: `AttributeError: 'Random' object has no attribute 'choices'. Did you mean: 'choice'?`.

Seems like this have been broken for sandboxes using Python 3.11.  The change on [`tempfile.py`](https://github.com/python/cpython/commit/d2810054c7ee1b4ce925fc520224b595b53bf4b4) was originally released on Python 3.10. The `random2` package [doesn't include the choices method](https://github.com/strichter/random2/blob/master/src/random2.py#L901).

I guess the main question would be: when can we start dropping the python2 compatibility layer (six, etc)?. Support for Python 2 was dropped since Lilac IIRC.

Reply:
We've been continuing to run Python 3.8 sandboxes and so we haven't run into that yet. Are you saying that the new code breaks on 3.11, or the existing prelude does?

Reply:
Author code that uses `choices()` or imports a module that uses it will fail.

Reply:
I'll add that we are planning to upgrade our sandboxes to 3.12 soon (within 1-2 months) as we switch to the enw codejail-service.

Reply:
I wonder how much of a pain it will be to change the prelude... I'm imagining it was mostly set up during the 2-to-3 migration, and there are probably a lot of xblocks still out there on edx.org and other deployments that still rely on them.

We might need to end up doing some kind of "edition" system where people can opt into newer versions of Python and/or the prelude, and older xblocks stay on older versions.

Reply:
IIRC, the reason we are still using `random2` is not for Python 2 _syntax_ compatibility, but specifically so that the results of random computations in course content do not change. For example, if a certain learner sees options `C,A,D,B` for a given `loncapa/python` problem, upgrading from `random2` to Python 3's `random` package might change that order to `B,D,C,A`. @feanil I remember that you dealt with this during the 2->3 upgrade; do I have the right

> We might need to end up doing some kind of "edition" system where people can opt into newer versions of Python and/or the prelude, and older xblocks stay on older versions.

This does seem like the right long-term solution.

Reply:
Dave's comment is particularly salient: https://github.com/openedx/edx-platform/pull/21893#discussion_r330791342